[Analysis] Bail out for negative offsets in isDereferenceableAndAlignedInLoop

david-arm opened the pull request:
This patch is effectively NFC because it doesn't actually affect any existing tests. It's not possible to even write a test for it, because by accident negative offsets are caught by a subsequent unsigned add overflow check. However, I think it's better to bail out explicitly for negative offsets so that it's more consistent with the unsigned remainder and add calculations.

fhahn commented:
Thanks for splitting this off!

I think https://llvm.godbolt.org/z/boYbjenvq may serve as a test case; it is using a pointer width of 16 (via the data layout and an alloca that's spanning more than half the available address space) and the GEP computing the address for the conditional load must wrap I think, which would make the `inbounds` GEP poison, so we shouldn't execute this unconditionally.

david-arm committed:
[LoopVectorize] Add negative offset tests for dereferenceable loop loads

david-arm committed:
[Analysis] Bail out for negative offsets in isDereferenceableAndAlignedInLoop

This patch is effectively NFC because it doesn't actually affect
any existing tests. It's not possible to even write a test for it,
because by accident negative offsets are caught by a subsequent
unsigned add overflow check. However, I think it's better to bail
out explicitly for negative offsets so that it's more consistent
with the unsigned remainder and add calculations.

david-arm commented:
> Thanks for splitting this off!
> 
> I think https://llvm.godbolt.org/z/boYbjenvq may serve as a test case; it is using a pointer width of 16 (via the data layout and an alloca that's spanning more than half the available address space) and the GEP computing the address for the conditional load must wrap I think, which would make the `inbounds` GEP poison, so we shouldn't execute this unconditionally.

Thanks a lot for this @fhahn! Your test has actually exposed a genuine bug that's now fixed by this patch. We were treating loads with negative offsets from allocas as being dereferenceable.

fhahn reviewed (approved):
LGTM, thanks!
